[Nah Choon Yan] iP

Choonyan02 opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/Choonyan02/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😄 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
___
If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Choonyan02 committed:
Added rename, greet and exit

Choonyan02 committed:
Echo

Choonyan02 committed:
Add, List

Choonyan02 committed:
Mark as Done

Choonyan02 committed:
ToDo, Event, Deadline

Choonyan02 committed:
Renamed Paimon.java to Duke.java

Choonyan02 committed:
Automated Text UI Testing

Choonyan02 committed:
Handle Errors

Choonyan02 committed:
Delete

Choonyan02 committed:
Add load and save for Level 7

Choonyan02 committed:
Merge branch 'branch-Level-7'

Choonyan02 committed:
Date Time format for Level-8

Choonyan02 committed:
Merge branch 'branch-Level-8'

Choonyan02 committed:
Made the code more OOP

Choonyan02 committed:
Divide classes into packages

Choonyan02 committed:
Merge branch 'add-gradle-support'

Choonyan02 committed:
Add JUint tests to test the behavior of the code

Choonyan02 committed:
Give users a way to find a task by searching for a keyword

Choonyan02 committed:
Add javadocs for Deadline and Task

Choonyan02 committed:
Add JavaDoc comments

Choonyan02 committed:
Merge branch 'branch-A-JavaDoc'

Choonyan02 committed:
Add JavaDocs for Test files

Choonyan02 committed:
fix bug in TaskList and also build gradle

yyyaohhh commented on src/test/java/duke/TodoTest.java:
```diff
@@ -0,0 +1,41 @@
+public class TodoTest {
+
+    /**
+     * Test the toString method for a not done task.
```
Good job for adding JavaDocs to all classes and methods, even for the test classes! 👍

yyyaohhh commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,123 @@
+    public static void greetings() {
+        showLine();
+        System.out.println(" Hello! I'm Paimon");
+        System.out.println(" What can I do for you?");
```
Perhaps customising these lines to the actual lines that Paimon would say in Genshin Impact 😉

yyyaohhh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,112 @@
+            try {
+                String taskDescription = command.substring(5);
+                if (taskDescription.length() == 0) {
+                    throw new DukeException("The description of a todo cannot be empty.");
```
Similar as the comment given for Ui, perhaps you can try to customise Paimon and report exceptions in a Paimon style 😀

Choonyan02 replied:
Thanks for the feedback, will try to do so!

yyyaohhh commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,48 @@
+package duke;
+
+import duke.task.Task;
+import java.io.*;
```
Perhaps you could list down all the specific ones that you have used, instead of `java.io.*`

Choonyan02 replied:
Resolved, thanks!

yyyaohhh commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,85 @@
+
+import duke.task.Task;
+import duke.task.TaskList;
+import static org.junit.jupiter.api.Assertions.*;
```
same as the above comment, try to list out all the specific ones you have used, instead of using `org.junit.jupiter.api.Assertions.*` to import all of them.

Choonyan02 replied:
Resolved, thanks!

yyyaohhh commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,85 @@
+        taskList.addTask(task);
+        assertEquals(task, taskList.getTask(initialTaskCount));
+    }
+}
```
Great job for creating many untrivial test cases!👍

yyyaohhh reviewed (commented):
Great job for the huge amount of effort put in, such as many test cases and JavaDocs written for all classes and methods even though is they are optional. Keep up the good work! (*^_^*)
Just some small things to take note, on importing specific classes you have used, instead of using `*`.
Look forward to seeing the cute customised Paimon!!♥

Choonyan02 committed:
Use checkstyle to detect coding violations

Choonyan02 committed:
Merge branch 'branch-A-CheckStyle'

Choonyan02 committed:
Make changes

Choonyan02 committed:
Add a GUI to the chatbot

Choonyan02 committed:
Merge branch 'branch-Level-10'

Darren159 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,109 @@
+     */
+    public static String readTask(String command, TaskList tasklist) {
+        String[] commandSplit = command.split(" ");
+        if (command.equals("bye")) {
+            System.exit(0);
+            return Ui.exit();
+        } else if (command.equals("list")) {
+            return Ui.listTasks(tasklist);
+        } else if (command.startsWith("find")) {
+            return tasklist.findTasks(commandSplit[1]);
+        } else {
+            if (command.startsWith("todo") || command.startsWith("deadline")
+                    || command.startsWith("event")) {
+                return addTask(command, tasklist);
+            } else if (command.startsWith("mark") || command.startsWith("unmark")
+                    || command.startsWith("delete")) {
+                try {
+                    if (commandSplit.length > 2 || Integer.parseInt(commandSplit[1]) > tasklist.taskCount()) {
+                        return Ui.showError(new DukeException("Please enter a valid number"));
+                    }
+                    int index = Integer.parseInt(commandSplit[1]) - 1;
+                    if (command.startsWith("mark")) {
+                        return tasklist.markTask(index);
+                    } else if (command.startsWith("unmark")) {
+                        return tasklist.unMarkTask(index);
+                    } else {
+                        return tasklist.deleteTask(index);
+                    }
+                } catch (NumberFormatException e) {
+                    return Ui.showError(new DukeException("Please enter a valid number"));
+                } catch (ArrayIndexOutOfBoundsException e) {
+                    return Ui.showError(new DukeException("Please enter a valid number"));
+                }
+            } else {
+                return Ui.showError(new DukeException("I'm sorry, but I don't know what that means :-("));
+            }
+        }
+    }
```
Try to avoid deep nesting as it makes the code harder to read. Maybe can consider doing the boolean checks in their own separate methods? Or can consider creating a Command package with all the individual commands?

Darren159 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,109 @@
+                String taskDescription = command.substring(5);
+                if (taskDescription.length() == 0) {
+                    return Ui.showError(new DukeException("The description of a todo cannot be empty."));
+                }
+                Task t = new Todo(taskDescription);
+                return tasklist.addTask(t);
+            } catch (StringIndexOutOfBoundsException e) {
+                return Ui.showError(new DukeException("The description of a todo cannot be empty."));
+            }
+        } else if (command.startsWith("deadline")) {
+            try {
+                String[] taskAndDeadline = command.substring(9).split(" /by ");
+                if (taskAndDeadline[0].length() == 0) {
+                    return Ui.showError(new DukeException("The description of a deadline cannot be empty."));
+                }
+                Task t = new Deadline(taskAndDeadline[0], taskAndDeadline[1]);
+                return tasklist.addTask(t);
+            } catch (StringIndexOutOfBoundsException e) {
+                return Ui.showError(new DukeException("The description of a deadline cannot be empty."));
+            } catch (ArrayIndexOutOfBoundsException e) {
+                return Ui.showError(new DukeException("The deadline (yyyy-MM-dd) is not specified."));
+            }
+        } else if (command.startsWith("event")) {
+            try {
+                String[] taskAndDate = command.substring(6).split(" /from | /to ");
+                if (taskAndDate[0].length() == 0) {
+                    return Ui.showError(new DukeException("The description of a todo cannot be empty."));
+                }
+                Task t = new Event(taskAndDate[0], taskAndDate[1], taskAndDate[2]);
+                return tasklist.addTask(t);
+            } catch (StringIndexOutOfBoundsException e) {
+                return Ui.showError(new DukeException("The description of an event cannot be empty."));
+            } catch (ArrayIndexOutOfBoundsException e) {
+                return Ui.showError(new DukeException("The dates (yyyy-MM-dd) of an event cannot be empty."));
+            }
+        } else {
+            return Ui.showError(new DukeException("I'm sorry, but I don't know what that means :-("));
+        }
+    }
+}
```
Same as above.

Darren159 reviewed (commented):
Overall code LGTM but I think adding a Command package would help make your Parser more readable! :)

Choonyan02 committed:
Improve code quality

Choonyan02 committed:
Use assertions

Choonyan02 committed:
Merge pull request #1 from Choonyan02/branch-A-Assertions

Use assertions

Choonyan02 committed:
Merge branch 'master' into branch-A-CodeQuality

Choonyan02 committed:
Merge pull request #2 from Choonyan02/branch-A-CodeQuality

Branch a code quality

Choonyan02 committed:
Add extention C-Help

Currently, there is no instruction for the user on how to use the chatbot

If the users do not know how to use the chatBot, they will feel frustrated

Add an extension C-Help, where user can find the commands available by entering "help"

Choonyan02 committed:
Add a user guide

Choonyan02 committed:
Update Event.java
